source/lac/petsc_vector_base.cc

```diff
@@ -259,8 +259,20 @@ namespace PETScWrappers
         AssertThrow(ierr == 0, ExcPETScError(ierr));
         ierr = VecGetArray(tvector, &array);
         AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+        // Store the indices we care about in the vector, so that we can then
+        // exchange this information between processes. It is unfortunate that
+        // we have to store integers in floating point numbers. Let's at least
+        // make sure we do that in a way that ensures that when we get these
+        // numbers back as integers later on, we get the same thing.
         for (PetscInt i = 0; i < end_index - ghost_start_index; i++)
-          array[i] = ghost_start_index + i;
+          {
+            Assert(static_cast<PetscInt>(static_cast<value_type>(
```

Comment:
I would prefer
```suggestion
            Assert(static_cast<PetscInt>(static_cast<PetscScalar>(
```
since `value_type` is only used from line 693 on. Alternatively, replace `PetscScalar` with `value_type` everywhere in this function.